File: UNIDAD 1/01_SOA_CON_BDD/00. SOAP_JAVA_EUREKABANK_GR10/02. CLICON/EUREKABANK_CLICON_GR10/src/ec/edu/monster/servicio/EurekaService.java
package ec.edu.monster.servicio;

import ec.edu.monster.ws.Movimiento;
import ec.edu.monster.ws.WSEureka;
import ec.edu.monster.ws.WSEureka_Service;
import java.util.List;

/**
 * Service class to interact with the WSEureka web service.
 * @author MATIAS
 */
public class EurekaService {

    private final WSEureka port;

    public EurekaService() {
        WSEureka_Service service = new WSEureka_Service();
        this.port = service.getWSEurekaPort();
    }

    public List<Movimiento> traerMovimientos(String cuenta) {
        return port.leerMovimientos(cuenta);
    }

    public int registrarDeposito(String cuenta, double importe, String codEmp) {
        return port.registrarDeposito(cuenta, importe, codEmp);
    }

    public int registrarRetiro(String cuenta, double importe, String codEmp) {
        return port.registrarRetiro(cuenta, importe, codEmp);
    }

    public int realizarTransferencia(String cuentaOrigen, String cuentaDestino, double importe, String codEmp) {
        return port.realizarTransferencia(cuentaOrigen, cuentaDestino, importe, codEmp);
    }

    public double verificarSaldo(String cuenta) {
        return port.verificarSaldo(cuenta);
    }

    public double obtenerCostoMovimiento(String cuenta) {
        return port.obtenerCostoMovimiento(cuenta);
    }
}

File: UNIDAD 1/00_SOA_SIN_BDD/002. REST_JAVA_CONUNI_GR10/02. CLIMOV/app/src/main/java/ec/edu/monster/servicio/EurekaService.kt
package ec.edu.monster.servicio

import com.google.gson.GsonBuilder
import ec.edu.monster.modelo.*
import ec.edu.monster.util.RestConstants
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

class EurekaService {
    private val api: EurekaApi

    init {
        val logging = HttpLoggingInterceptor().apply { level = HttpLoggingInterceptor.Level.BODY }
        val client = OkHttpClient.Builder().addInterceptor(logging).build()
        val gson = GsonBuilder().setDateFormat("yyyy-MM-dd'Z'").create()
        val retrofit = Retrofit.Builder()
            .baseUrl(RestConstants.BASE_URL)
            .addConverterFactory(GsonConverterFactory.create(gson))
            .client(client)
            .build()
        api = retrofit.create(EurekaApi::class.java)
    }

    suspend fun login(usuario: String, clave: String): Usuario? {
        val loginRequest = LoginRequest(usuario, clave)
        val response = api.login(loginRequest)
        return if (response.isSuccessful) response.body() else null
    }

    suspend fun traerMovimientos(cuenta: String): List<Movimiento> {
        val response = api.traerMovimientos(cuenta)
        if (response.isSuccessful) {
            return response.body() ?: emptyList()
        } else {
            throw Exception(response.errorBody()?.string() ?: "Error al consultar movimientos")
        }
    }

    suspend fun registrarDeposito(cuenta: String, importe: Double, codEmp: String) {
        val request = DepositoRequest(cuenta, importe, codEmp)
        val response = api.registrarDeposito(request)
        if (!response.isSuccessful) {
            throw Exception(response.errorBody()?.string() ?: "Error al registrar depósito")
        }
    }

    suspend fun registrarRetiro(cuenta: String, importe: Double, codEmp: String) {
        val request = RetiroRequest(cuenta, importe, codEmp)
        val response = api.registrarRetiro(request)
        if (!response.isSuccessful) {
            throw Exception(response.errorBody()?.string() ?: "Error al registrar retiro")
        }
    }

    suspend fun realizarTransferencia(cuentaOrigen: String, cuentaDestino: String, importe: Double, codEmp: String) {
        val request = TransferenciaRequest(cuentaOrigen, cuentaDestino, importe, codEmp)
        val response = api.realizarTransferencia(request)
        if (!response.isSuccessful) {
            throw Exception(response.errorBody()?.string() ?: "Error al realizar transferencia")
        }
    }

    suspend fun verificarSaldo(cuenta: String): Double {
        val response = api.verificarSaldo(cuenta)
        if (response.isSuccessful) {
            return response.body()?.saldo ?: -1.0
        } else {
            throw Exception(response.errorBody()?.string() ?: "Error al verificar saldo")
        }
    }

    suspend fun obtenerCostoMovimiento(cuenta: String): Double {
        val response = api.obtenerCostoMovimiento(cuenta)
        if (response.isSuccessful) {
            return response.body()?.costo ?: -1.0
        } else {
            throw Exception(response.errorBody()?.string() ?: "Error al obtener costo")
        }
    }
}